src/arithmetic_impls.rs

```diff
@@ -0,0 +1,294 @@
+use crate::{decimal::CalculationResult, ops, Decimal};
+use core::ops::{Add, Div, Mul, Rem, Sub};
+use num_traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedRem, CheckedSub};
+
+// Macros and `Decimal` implementations
+
+macro_rules! impl_checked {
+    ($(#[$($doc:meta)*])* $fun:ident, $impl:ident) => {
+        /// Like the checked version but returns lower or upper bounds instead of `None`.
+        $(#[$($doc)*])*
+        #[inline(always)]
+        #[must_use]
+        pub fn $fun(self, other: Decimal) -> Option<Decimal> {
+            match ops::$impl(&self, &other) {
+                CalculationResult::Ok(result) => Some(result),
+                _ => None,
+            }
+        }
+    };
+}
+
+macro_rules! impl_checked_and_saturating {
+    (
+        $(#[$($checked_doc:meta)*])*
+        $checked_fun:ident,
+        $checked_impl:ident,
+
+        $(#[$($saturating_doc:meta)*])*
+        $saturating_fun:ident
+    ) => {
+        impl_checked!($(#[$($checked_doc)*])* $checked_fun, $checked_impl);
+        impl_saturating!($(#[$($saturating_doc)*])* $saturating_fun, $checked_fun);
+    };
+}
+
+macro_rules! impl_saturating {
+    ($(#[$($doc:meta)*])* $fun:ident, $impl:ident) => {
+        $(#[$($doc)*])*
+        #[inline(always)]
+        #[must_use]
+        pub fn $fun(self, other: Decimal) -> Decimal {
+            if let Some(elem) = self.$impl(other) {
+                elem
+            }
+            else {
+                if self.is_sign_positive() {
+                    Decimal::MAX
+                }
+                else {
+                    Decimal::MIN
+                }
+            }
+        }
+    };
+}
+
+impl Decimal {
+    impl_checked_and_saturating!(
+        /// Checked addition. Computes `self + other`, returning `None` if overflow occurred.
+        ///
+        /// ```rust
+        /// use rust_decimal::Decimal;
+        ///
+        /// assert_eq!(Decimal::MAX.checked_add(Decimal::MAX), None);
+        /// assert_eq!(Decimal::MIN.checked_add(Decimal::MIN), None);
+        /// ```
```

Comment:
These macros can be de-duplicated in the near future using Rust 1.54 -> https://blog.rust-lang.org/2021/07/29/Rust-1.54.0.html

Reply:
I'm almost wondering if we upgrade to a minimum rust version of `1.54` preemptively? You're right - it's just around the corner, though I do think it'd help make these macros a little more readable. Admittedly, on first glance I thought the saturating alternative was missing documentation when it is actually present in the macro (or should be).

Reply:
Well, it is up to you but an early 1.54 would be very handy.

FWICT, the Rust team does not back-port security bugs to old stable versions. For example, from https://www.cvedetails.com/vulnerability-list/vendor_id-19029/product_id-48677/Rust-lang-Rust.html it is possible to see that https://www.cvedetails.com/cve/CVE-2018-1000810/ is present from 1.26 to 1.29 but only 1.29.1 onwards has a fixing patch.